File: npm/mind-expander/bin/install-skill.js
//
// `mind-expander install-skill` — detects which AI coding agents the
// user has installed, then copies the bundled skill file into the
// right path for each. Idempotent: re-running updates the installed
// copy in place (and replaces, not duplicates, any prior install).
//
// The skill file itself is the same `skill/mind-expander.md` that
// ships in this npm package via the `prepack` script in package.json.
//
// Per-agent install paths:
//   Claude Code  → ~/.claude/skills/mind-expander.md   (flat file)
//   Codex        → ~/.codex/AGENTS.md                  (marker-fenced append)
//   Cursor       → .cursor/rules/mind-expander.md      (project-level)
//
// The Claude Code path is user-level (works across all projects) since
// most users will want mind-expander available globally. Cursor's
// `.cursor/rules/` is project-local because Cursor doesn't have a
// clean user-level skills concept. Codex's AGENTS.md is user-level
// because Codex looks there for global instructions; the file may
// already exist with the user's own content, so we fence our block
// with HTML comment markers so subsequent re-installs replace cleanly.

'use strict';

const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');
const readline = require('node:readline');

// The bundled skill file. In a published install, prepack copies the
// repo's `skill/mind-expander.md` into `<pkg>/skill/`. In a source
// checkout (before prepack runs) it still lives at the repo root,
// two levels up from this script. Try both so the command works
// either way — useful for smoke-testing without a publish round-trip.
const SKILL_CANDIDATES = [
  path.join(__dirname, '..', 'skill', 'mind-expander.md'),
  path.join(__dirname, '..', '..', '..', 'skill', 'mind-expander.md'),
];
const SKILL_SOURCE =
  SKILL_CANDIDATES.find((p) => fs.existsSync(p)) ?? SKILL_CANDIDATES[0];

// Each target describes one agent we know how to install to. `detect`
// returns whether the agent appears to be present on this machine;
// `install` does the actual copy/write and returns the path written.
function buildTargets() {
  const home = os.homedir();
  const cwd = process.cwd();
  return [
    {
      name: 'Claude Code',
      scope: 'user',
      detect: () => fs.existsSync(path.join(home, '.claude')),
      describe: () => path.join(home, '.claude', 'skills', 'mind-expander.md'),
      install: (skill) => {
        const dir = path.join(home, '.claude', 'skills');
        const dest = path.join(dir, 'mind-expander.md');
        fs.mkdirSync(dir, { recursive: true });
        fs.writeFileSync(dest, skill);
        return dest;
      },
    },
    {
      name: 'Codex',
      scope: 'user',
      detect: () => fs.existsSync(path.join(home, '.codex')),
      describe: () => path.join(home, '.codex', 'AGENTS.md') + ' (appended)',
      install: (skill) => {
        // Codex's AGENTS.md is a single combined-instructions file.
        // We append our block with HTML-comment markers so any prior
        // install can be cleanly replaced on re-run without
        // duplicating, and so users can hand-edit around our block
        // without our updates clobbering theirs.
        const dest = path.join(home, '.codex', 'AGENTS.md');
        const START = '<!-- mind-expander:start -->';
        const END = '<!-- mind-expander:end -->';
        let existing = '';
        if (fs.existsSync(dest)) {
          existing = fs.readFileSync(dest, 'utf8');
          const s = existing.indexOf(START);
          const e = existing.indexOf(END);
          if (s !== -1 && e !== -1 && e > s) {
            existing =
              existing.slice(0, s).trimEnd() +
              existing.slice(e + END.length);
          }
        }
        const block = `\n\n${START}\n${skill.trim()}\n${END}\n`;
        fs.mkdirSync(path.dirname(dest), { recursive: true });
        fs.writeFileSync(dest, existing.trimEnd() + block);
        return dest;
      },
    },
    {
      name: 'Cursor',
      scope: 'project',
      detect: () =>
        fs.existsSync(path.join(cwd, '.cursor')) ||
        fs.existsSync(path.join(cwd, '.cursorrules')),
      describe: () => path.join(cwd, '.cursor', 'rules', 'mind-expander.md'),
      install: (skill) => {
        const dir = path.join(cwd, '.cursor', 'rules');
        const dest = path.join(dir, 'mind-expander.md');
        fs.mkdirSync(dir, { recursive: true });
        fs.writeFileSync(dest, skill);
        return dest;
      },
    },
  ];
}

function ask(prompt) {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  return new Promise((resolve) => {
    rl.question(prompt, (ans) => {
      rl.close();
      resolve(ans.trim().toLowerCase());
    });
  });
}

async function main() {
  if (!fs.existsSync(SKILL_SOURCE)) {
    // Would only happen if someone ran us out of a broken/stripped
    // install where prepack didn't bundle the skill file.
    process.stderr.write(
      `install-skill: bundled skill file not found at ${SKILL_SOURCE}\n` +
        `This is a packaging bug — please file an issue.\n`,
    );
    process.exit(1);
  }
  const skill = fs.readFileSync(SKILL_SOURCE, 'utf8');

  const targets = buildTargets();
  const detected = targets.filter((t) => t.detect());

  if (detected.length === 0) {
    process.stdout.write(
      `mind-expander install-skill\n\n` +
        `No supported agents detected on this machine.\n` +
        `Looked for: ~/.claude/, ~/.codex/, .cursor/ (in current directory).\n\n` +
        `If you have an agent installed in a non-standard location, copy\n` +
        `the bundled skill file manually:\n\n` +
        `  ${SKILL_SOURCE}\n`,
    );
    return;
  }

  process.stdout.write(
    `mind-expander install-skill\n\n` +
      `Detected ${detected.length} agent${detected.length === 1 ? '' : 's'}:\n`,
  );
  for (const t of detected) {
    process.stdout.write(`  • ${t.name} (${t.scope}) → ${t.describe()}\n`);
  }
  process.stdout.write('\n');

  for (const t of detected) {
    const ans = await ask(`Install skill to ${t.name}? [Y/n] `);
    if (ans !== '' && ans !== 'y' && ans !== 'yes') {
      process.stdout.write(`  skipped\n\n`);
      continue;
    }
    try {
      const dest = t.install(skill);
      process.stdout.write(`  ✓ wrote ${dest}\n\n`);
    } catch (err) {
      process.stderr.write(`  ✗ failed: ${err.message}\n\n`);
    }
  }

  process.stdout.write(
    `Done. Re-run \`npx mind-expander install-skill\` after upgrading\n` +
      `the package to refresh the installed skill file.\n`,
  );
}

module.exports = main;

File: src/install_skill.rs
use std::fs;
use std::io::{self, BufRead, Write};
use std::path::{Path, PathBuf};

const SKILL: &str = include_str!("../skill/mind-expander.md");

struct Agent {
    name: &'static str,
    scope: &'static str,
    dest: PathBuf,
}

pub fn run(yes: bool) -> anyhow::Result<()> {
    let home =
        dirs::home_dir().ok_or_else(|| anyhow::anyhow!("cannot determine home directory"))?;
    let cwd = std::env::current_dir()?;

    let mut agents: Vec<Agent> = Vec::new();

    if home.join(".claude").is_dir() {
        agents.push(Agent {
            name: "Claude Code",
            scope: "user",
            dest: home
                .join(".claude")
                .join("skills")
                .join("mind-expander")
                .join("SKILL.md"),
        });
    }
    if home.join(".codex").is_dir() {
        agents.push(Agent {
            name: "Codex",
            scope: "user",
            dest: home
                .join(".agents")
                .join("skills")
                .join("mind-expander")
                .join("SKILL.md"),
        });
    }
    if cwd.join(".cursor").is_dir() || cwd.join(".cursorrules").exists() {
        agents.push(Agent {
            name: "Cursor",
            scope: "project",
            dest: cwd.join(".cursor").join("rules").join("mind-expander.mdc"),
        });
    }

    if agents.is_empty() {
        println!("mind-expander install-skill\n");
        println!("No supported agents detected on this machine.");
        println!("Looked for:");
        println!("  \u{2022} ~/.claude/  (Claude Code)");
        println!("  \u{2022} ~/.codex/   (Codex)");
        println!("  \u{2022} .cursor/    (Cursor, in current directory)\n");
        println!("Copy the bundled skill file manually if needed:\n");
        println!("  Claude Code:  ~/.claude/skills/mind-expander/SKILL.md");
        println!("  Codex:        ~/.agents/skills/mind-expander/SKILL.md");
        println!("  Cursor:       <project>/.cursor/rules/mind-expander.mdc");
        return Ok(());
    }

    println!("mind-expander install-skill\n");
    println!(
        "Detected {} agent{}:",
        agents.len(),
        if agents.len() == 1 { "" } else { "s" }
    );
    for a in &agents {
        println!(
            "  \u{2022} {} ({}) \u{2192} {}",
            a.name,
            a.scope,
            a.dest.display()
        );
    }
    println!();

    for a in &agents {
        if !yes {
            print!("Install skill to {}? [Y/n] ", a.name);
            io::stdout().flush()?;
            let mut line = String::new();
            io::stdin().lock().read_line(&mut line)?;
            let ans = line.trim().to_lowercase();
            if !ans.is_empty() && ans != "y" && ans != "yes" {
                println!("  skipped\n");
                continue;
            }
        }

        let cleaned = cleanup_v012(a, &home, &cwd);

        if let Some(parent) = a.dest.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::write(&a.dest, SKILL)?;
        println!("  \u{2713} wrote {}", a.dest.display());
        if cleaned {
            println!("    (also cleaned up the broken v0.1.2 install)");
        }
        println!();
    }

    println!("Done. Re-run `mind-expander install-skill` after upgrading");
    println!("the package to refresh the installed skill files.");
    Ok(())
}

fn cleanup_v012(agent: &Agent, home: &Path, cwd: &Path) -> bool {
    match agent.name {
        "Claude Code" => {
            let stale = home.join(".claude").join("skills").join("mind-expander.md");
            try_remove(&stale)
        }
        "Codex" => {
            let agents_md = home.join(".codex").join("AGENTS.md");
            cleanup_codex_agents_md(&agents_md)
        }
        "Cursor" => {
            let stale = cwd.join(".cursor").join("rules").join("mind-expander.md");
            try_remove(&stale)
        }
        _ => false,
    }
}

fn try_remove(path: &Path) -> bool {
    matches!(fs::remove_file(path), Ok(()))
}

fn cleanup_codex_agents_md(path: &Path) -> bool {
    let content = match fs::read_to_string(path) {
        Ok(c) => c,
        Err(_) => return false,
    };
    const START: &str = "<!-- mind-expander:start -->";
    const END: &str = "<!-- mind-expander:end -->";
    let s = match content.find(START) {
        Some(i) => i,
        None => return false,
    };
    let e = match content.find(END) {
        Some(i) => i,
        None => return false,
    };
    if e <= s {
        return false;
    }
    let before = content[..s].trim_end();
    let after = content[e + END.len()..].trim_start();
    if before.is_empty() && after.is_empty() {
        let _ = fs::remove_file(path);
    } else {
        let stripped = match (before.is_empty(), after.is_empty()) {
            (true, _) => format!("{after}\n"),
            (_, true) => format!("{before}\n"),
            _ => format!("{before}\n\n{after}\n"),
        };
        let _ = fs::write(path, stripped);
    }
    true
}
